rfcs/0019-storing-client-data.md

```diff
@@ -0,0 +1,445 @@
+
+For the objects discussed in the previous section, we can embed them as tables in one global SQL database. Below are the schemas of the tables:
+
+```
+Table public_keys {
+  id serial [pk]
+  value text [not null]
+  Indexes {
+    value [name: "public_key"]
+  }
+}
+
+Table blocks {
+  id serial [pk]
+  state_hash string [not null]
+  coinbase int [not null]
+  creator string [not null]
+  staged_ledger_hash string [not null]
+  ledger_hash string [not null]
+  global_slot int [not null] // encompasses (epoch, slot)
+  ledger_proof_nonce int [not null]
+  status int [not null]
+  block_length int [not null]
+  block_time bit(64) [not null]
+  Indexes {
+    state_hash [name: "state_hash"]
+    (block_length, global_slot) [name: "block_compare"]
+    block_time [name: "block_compare"]
+  }
+}
+
+  // Consensus_status is an ADT, but with its small state spaces, it can be coerced into an integer to make it easy to store it in a database. Here are the following states:
+  // - pending : The block is in the transition frontier and its value is its block confirmation number. Value[0, (k-1)]
+  // - confirmed : The block has reached finality by passing the root. Value=-1
+  // - failure: The block got removed from the transition frontier. Value=-2
+  // - unknown: After becoming online and bootstrapping, we may not know what will happen to the block. Value=-3
+
+Enum user_command_type {
+  payment
+  delegation
+}
+
+Table user_commands {
+  id serial [pk]
+  hash string [not null, unique]
+  type user_command_type [not null]
+  nonce bit(64) [not null]
+  sender int [not null, ref: > public_keys.id]
+  receiver int [not null, ref: > public_keys.id]
+  amount bit(64) [not null]
+  fee bit(64) [not null]
+  memo string [not null]
+  first_seen bit(64)
+  Indexes {
+    hash [name: "user_command_hash"]
+    sender [name: "user_command_sender"]
+    receiver [name: "user_command_receiver"]
+    (sender, first_seen) [name: "fast_sender_pagination"]
+    (receiver, first_seen) [name: "fast_receiver_pagination"]
+  }
+}
+
+Table fee_transfers {
+  id serial [pk]
+  hash string [not null, unique]
+  fee bit(64) [not null]
+  receiver int [not null, ref: > public_keys.id]
+  first_seen bit(64)
+  Indexes {
+    hash [name: "fee_transfers_hash"]
+    receiver [name:"fee_transfer_receiver"]
+  }
+}
+
+Table receipt_chain_hashes {
+  id serial [pk]
+  hash string [not null, unique]
+  previous_hash string
+  user_command_id string [not null, ref: > user_commands.id]
+  Indexes {
+    user_command_id
+  }
+}
+
+Table block_to_user_commands {
+  block_id int [not null, ref: > blocks.id]
+  transaction_id int [not null, ref: > user_commands.id]
+  receipt_chain_id int [ref: > receipt_chain_hashes.id]
+  Indexes {
+    block_id [name:"block_to_user_command.block"]
+    transaction_id [name: "block_to_user_command.transaction"]
+  }
```

Comment:
IMO, we should not specify indices for our join tables yet. Indices on join tables are not free (as all indices are not), and the way you want to set them up on join tables is very dependent on the queries you are planning on doing on that table. Indices are always easy to add later. I generally err on the side of defining very few indices initially and adding them to optimize later. It's easier to identify which indices you need later than it is to identify which ones you don't.

Reply:
I can omit the indices but I am highly confident that we would make a lot queries that uses these join tables as look ups, like the blocks query to get the transactions in a block. If we do not have these indices for the join tables, these lookups are going to be slow.

Usually indices will impact write performance. However, our system does not require us to make a lot of writes. The write transaction that has the most amount of latency is writing a block. When we launch mainnet, we would probably receive at most 30 blocks per slot and the slot time would be at the minimum 30 seconds (given that we optimize the heck out of the protocol). Writing a block transaction to the database would only be milliseconds with this database. I don't think these indices on the join tables would kill performance given our system requirements.